### DISCUSSION TITLE
Example of the generator with yield #question

### DISCUSSION THREAD

**Discussion by inv2004** [PyO3/pyo3#2985] [Questions]
Hello,

Is it possible to call some python function which uses generator (would like to create it in rust part).
What is correct way to do yield(val) from pyo3 side?

Im afraid that calling from python could affect performance significantly.
Do you have any examples (with PyIterator?)?

Thank you,

**Comment by konstin** [comment]
I don't think calling `yield` in rust is what you want; yield is just a syntactic sugar that will build an object with `__iter__` and `__next__` methods from that function. What you can do in rust is to create an object that has `__iter__` and `__next__` which you can pass to your python code.

**Comment by inv2004** [comment]
Unfortunately, stuck a bit:

I found one example, but it uses yield:

https://pyo3.rs/master/doc/src/pyo3/types/iterator.rs.html#157

**Comment by konstin** [comment]
I'm not really sure what you're trying to do exactly, but I suppose you could try the following

https://github.com/PyO3/pyo3/blob/52dfc0013c81e1372e831fa5087656dff4e26dcf/tests/test_dunder.rs#L52-L67

This will allow returning your struct to python where it can be used just like a generator.

**Comment by inv2004** [comment]
Works perfect. Thank you.

**Comment by jymchng** [comment]
Hi

Just curious if the object I wish to 'yield' is not a struct but a `#[pyfunction]`, how should I be writing this?

**Reply by mejrs** [reply]
Can you clarify what you mean? This question doesn't really make sense.

**Reply by jymchng** [reply]
Sorry, I should have been clearer.

In python, I want to be able to do the following:

```
class CustomCanBeValidatedType:
    @classmethod
    def __get_validators__(cls):
       yield cls.validate

    @classmethod
    def validate(cls, v, values):
        # do some validations
        return v
```

Here, the PyObject that is being yielded is a PyFunction, is there a way to obtain this behavior in Rust through pyo3?

Thank you.

**Reply by mejrs** [reply]
Returning functions is not related to using yield. Your best option is to make an iterator.
